vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs

```diff
@@ -115,51 +93,130 @@ type private FSharpProjectOptionsReactor (workspace: VisualStudioWorkspaceImpl,
+    /// We don't just clear the cps command line options as a callback could have occured from HandleCommandLineChanges that is for the new solution.
+    let cleanupCpsCommandLineOptions (solution: Solution) =
+        let projectIds = cpsCommandLineOptions.Keys |> Array.ofSeq
+        for projectId in projectIds do
+            if not (solution.ContainsProject projectId) then
+                cpsCommandLineOptions.TryRemove projectId |> ignore
+
+    let setSolution (solution: Solution) =
+        match currentSolution with
+        | Some oldSolution when solution.Id <> oldSolution.Id ->
+            cache.Clear ()
+            checkerProvider.Checker.StopBackgroundCompile ()
+            checkerProvider.Checker.InvalidateAll ()
+            cleanupCpsCommandLineOptions solution
+
+        | Some oldSolution when solution.Version <> oldSolution.Version ->
+            checkerProvider.Checker.StopBackgroundCompile ()
+            let changes = solution.GetChanges oldSolution
+            for removedProject in changes.GetRemovedProjects() do
+                invalidateProject removedProject.Id
+            cleanupCpsCommandLineOptions solution
+
+        | _ -> ()
+
+        currentSolution <- Some solution
+
+    let checkProject (oldProject: Project) (newProject: Project) (settings: EditorOptions) ct =
+        async {
+            Debug.Assert (oldProject.Id = newProject.Id)
+
+            if oldProject.Version <> newProject.Version then
+                invalidateProjectCache oldProject.Id
+                return false
+            elif settings.LanguageServicePerformance.EnableInMemoryCrossProjectReferences then
+                let! oldVersion = oldProject.GetDependentVersionAsync ct |> Async.AwaitTask
+                let! newVersion = newProject.GetDependentVersionAsync ct |> Async.AwaitTask
+
+                if oldVersion <> newVersion then
+                    // invalidate any project that depends on this project
+                    newProject.Solution.GetProjectDependencyGraph().GetProjectsThatTransitivelyDependOnThisProject newProject.Id
+                    |> Seq.iter (fun projectId ->
+                        invalidateProjectCache projectId
+                    )
+
+                    // while we are not invalidated, we should update the project itself in the cache.
+                    updateProject newProject
+                    return true
+                else
+                    return true
+            else
+                return true
+        }
 
     let rec tryComputeOptionsByFile (document: Document) (ct: CancellationToken) =
         async {
-            let! fileStamp = document.GetTextVersionAsync(ct) |> Async.AwaitTask
-            match singleFileCache.TryGetValue(document.Id) with
+            let isScript = isScriptFile document.FilePath
+            let! version = document.Project.GetDependentVersionAsync ct |> Async.AwaitTask
```

Comment:
AFAICS `version` is not used in the major branch of the `match` so it should be pushed into that branch?

Reply:
Yup, good catch, I can move it.